### PR TITLE
[Refactor] XContentFactory contentType introspection to MediaTypeRegistry

### PR DESCRIPTION
**Pull request by nknize**
`XContentFactory` is tightly coupled to concrete `XContentType`. This PR builds on the `MediaType` abstractions, specifically `MediaTypeRegistry`, to decouple `contentType(` introspection (e.g., determining type from byte streams) from concrete `XContentType` implementations. This  removes the tight coupling of `:libs:opensearch-x-content` with abstract interface contracts currently defined in `:server` further enabling the ability to refactor those contracts from `:server` to the proper library for modularity and extensibility.

relates #5910
relates #8110 

depends on #8636

### CODE REVIEW

**Comment by nknize** [comment]
Reminder: this PR is larger than it's actual change because it includes (depends on) #8636

**Comment by nknize** [comment]
For posterity I ran across a flaky test w/ `RemoteFSTranslogTests` when running `:server:test` locally. It does not repro on it's own (hence "flaky") so I could not chase the failure. 

```
./gradlew ':server:test' --tests "org.opensearch.index.translog.RemoteFSTranslogTests.testConcurrentWriteViewsAndSnapshot" -Dtests.seed=11209ED8456F2E6A -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=sr-ME -Dtests.timezone=Pacific/Johnston -Druntime.java=20
```
I documented this on # 5823.

File: libs/core/src/main/java/org/opensearch/core/xcontent/MediaTypeParserRegistry.java [hunk 1]
```diff
@@ -93,6 +101,213 @@ public static MediaType fromFormat(String format) {
         return formatToMediaType.get(format.toLowerCase(Locale.ROOT));
     }
 
+    /**
+     * Returns a binary content builder for the provided content type.
+     */
+    public static XContentBuilder contentBuilder(MediaType type) throws IOException {
```
[the remark this answers is not in the document]
Comment (nknize) [reply libs/core/src/main/java/org/opensearch/core/xcontent/MediaTypeParserRegistry.java]: Yeah. I took care of that.

**Comment by nknize** [comment]
I switched this to use SPI instead of trying to juggle the classloading logic. When the `MediaTypeRegistry` is first loaded it will use SPI to load all MediaTypes through the `MediaTypeProvider`. I think this is far cleaner and will avoid race conditions. I'll update the commit message and the PR description.

**Comment by nknize** [comment]
Oh how annoying. Instead of git recognizing MediaTypeRegistry as a rename it made it a new file. :/  oye...

**Comment by nknize** [comment]
welp!!! Github seems to have a low similarity index (probably 50%? anyone know?) Since `MediaTypeRegistry` content has significant method adds github detects it as a new file. I don't think it's a big issue, though. So I'm going to stop putting time into incrementally changing it.

**Comment by nknize** [comment]
@reta  If I refactor to a new file in this commit github's similarity index won't detect it as a rename. I spent quite a bit of time trying to adjust that similarity index without luck. Unless you know how to change it I'd rather retain history than refactor the classname in this PR. So I reverted the rename of `MediaTypeParserRegistry` and will do that in a follow up PR.

File: libs/core/src/main/java/org/opensearch/core/xcontent/MediaTypeParserRegistry.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (nknize) [reply libs/core/src/main/java/org/opensearch/core/xcontent/MediaTypeParserRegistry.java:131]: I reverted in this PR and will do it in a followup. See [this comment](https://github.com/opensearch-project/OpenSearch/pull/8826#issuecomment-1654627608)

**Comment by nknize** [comment]
To help tease apart some of these changes I went ahead and opened two incremental PRs. Lets merge those first in this order:

1. #8940 - Rote refactor the `MediaTypeParserRegistry` name to just `MediaTypeRegistry` per https://github.com/opensearch-project/OpenSearch/pull/8826#discussion_r1273746991 
2. #8938 - add the SPI code so we don't conflate this with an enhancement mechanism. 
 
Once those two PRs are merged I'll rebase this PR to refactor the `contentType` introspection logic in this PR.

File: libs/core/src/main/java/org/opensearch/core/xcontent/MediaTypeParserRegistry.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (nknize) [reply libs/core/src/main/java/org/opensearch/core/xcontent/MediaTypeParserRegistry.java:131]: Update: check my latest [comment](https://github.com/opensearch-project/OpenSearch/pull/8826#issuecomment-1654688903) for two PRs that can merge before this change.

**Comment by nknize** [comment]
@reta this one will be ready to go as well

**Review by reta — APPROVED** [review]

**Pull request merged**